### DISCUSSION TITLE
Repeated module calls in terragrunt

### DISCUSSION THREAD

**Discussion by YR-ZR0** [gruntwork-io/knowledge-base#711] [How do I]
Hi All,

I'm currently trying to use terragrunt with the following module:

https://github.com/terraform-aws-modules/terraform-aws-iam/tree/master/modules/iam-role-for-service-accounts-eks

The part I am looking for guidance on is DRY'ing up repeated module calls. Mostly in reference to the file here:

https://github.com/terraform-aws-modules/terraform-aws-iam/blob/master/examples/iam-role-for-service-accounts-eks/main.tf

What is the best way to deal with a module where structure is repeated to create unique resources ?

Thanks

---

<ins datetime="2023-05-15T16:31:40Z">
  <p><a href="https://support.gruntwork.io/hc/requests/110166">Tracked in ticket #110166</a></p>
</ins>

**Comment by gitsstewart** [comment]
@YR-ZR0 have you already taken a look at our Terragrunt docs specific to overall keeping your code DRY? https://terragrunt.gruntwork.io/docs/features/keep-your-terraform-code-dry/

**Comment by YR-ZR0** [comment]
Hi @gitsstewart thanks for getting back to me, I've currently got the following in an attempt to keep things DRY and hand off as much to terragrunt as possible.

```hcl
generate "vpc_cni_irsa_role" {
  path      = "vpc_cni.tf"
  if_exists = "overwrite_terragrunt"
  contents  = <<EOF
data "aws_partition" "current" {}
locals {
  partition = data.aws_partition.current
}

module "vpc_cni_irsa_role" {
  source = "terraform-aws-modules/iam/aws//modules/iam-role-for-service-accounts-eks"

  role_name = "vpc-cni"

  attach_vpc_cni_policy = true
  vpc_cni_enable_ipv4   = true

  oidc_providers = {
    main = {
      provider_arn               = "${dependency.eks.outputs.oidc_provider_arn}"
      namespace_service_accounts = ["kube-system:aws-node"]
    }
  }
}
EOF

}
```

The main issue i have here is that the "role_name" parameter is unique to each call of the module if you are creating many roles. And from my understanding we can't use a for_each to loop a module.

Hope the code snippet helps in understanding the issue

**Reply by denis256** [reply]
Hi,
I was thinking about extraction of module as dependency and pass required inputs :
```
├── app
│   ├── main.tf
│   └── terragrunt.hcl
├── dep
│   ├── dep1
│   │   └── terragrunt.hcl
│   └── dep2
│       └── terragrunt.hcl
├── tf-module
│   └── main.tf
└── tf_module.hcl
```

```
# app/terragrunt.hcl 
dependencies {
  paths = ["../dep/dep1", "../dep/dep2"]
}

# dep/dep1/terragrunt.hcl 

include "tf_module" {
  # include terraform module configuration: repo/path
  path = find_in_parent_folders("tf_module.hcl")
}

# custom inputs
inputs = {
  file_name = "file1.txt"
  content  = "content of file1.txt"
}


```
